### ISSUE TITLE
autocomplete: problems w/ option navigation on iOS VoiceOver and JAWS w/ IE11

### ISSUE THREAD

**Issue by jelbourn** [angular/material#10970]
Seems to be no way to select an option from the list

**Comment by benjiallen** [comment]
Adding more details:

* Component with issue
    - `md-autocomplete`
* Overview of the Issue
    - Given I am using JAWS and I am interacting with the `md-autocomplete` component, when I enter a character that has search results, then JAWS will announce that there are matches available but there is no way to select a search result
        + You can hear the first search result but you cannot select it
    - Given I am using iOS VoiceOver and I am interacting with the `md-autocomplete` component, when I enter a character that has search results, then JAWS will announce that there are matches available but there is no way to select a search result
        + As soon as you try to select a search result using the "swipe left to right" gesture the search results will disappear
* AngularJS Material Version
    - 1.1.5
* Browsers and Operating System
    - iOS 11.2.1, Safari, VoiceOver
    - Windows 7, IE11, JAWS 17.0.2619
* Reproduce the Error
    - Visit https://material.angularjs.org/latest/demo/autocomplete and interact with the "Basic Usage" component with either JAWS or VoiceOver

**Comment by Splaktar** [comment]
I get the following on VoiceOver for macOS:
![screen shot 2018-08-11 at 4 22 43 pm](https://user-images.githubusercontent.com/3506071/43995775-01f32e7c-9d83-11e8-80c7-ed1c7ade8368.png)
This indeed doesn't seem to give the proper instructions for navigating the options (up/down arrows) and it doesn't say to press enter to select an option.

On Chromevox, I get the following:
"What is your favorite US state? Expanded Combo box Form is editing  Press up or down arrow for auto completions."
This is more helpful, but it doesn't say to press enter to select an option.

Both of these seem to be using the default guidance by the screen readers for `role="combobox"`.

The [WAI-ARIA Authoring Practices for Combobox](https://www.w3.org/TR/wai-aria-practices/#combobox) describe that the up/down arrows should be used for selecting options. It also details the behavior of the Enter key, which we follow.

@benjiallen is this issue a matter of discoverability of how to interact with the component or is it a bug with the up/down/enter keys not working on iOS VoiceOver and JAWS with IE11?

I'm not familiar with the to select a search result using the "swipe left to right" in iOS VoiceOver, do you have a link the explains it?

**Comment by Splaktar** [comment]
This may be fixed by https://github.com/angular/material/pull/11441. I need someone to test on JAWS and iOS VoiceOver to verify this as I don't have access to JAWS or an iOS device. 

For the time being, this can be tested via https://angularjs-material.firebaseapp.com/ or at https://material.angularjs.org/HEAD/ once PR https://github.com/angular/material/pull/11441 is merged.

**Comment by Splaktar** [comment]
@benjiallen thank you very much for the information! I am a little more familiar with those gestures now after using Talkback for Android.

However, I don't have an iOS device and I don't have a license for JAWS, so I would really appreciate your help with testing these latest changes to go from WAI-ARIA 1.0 to WAI-ARIA 1.1 with autocomplete. You can find examples using 1.1 here: https://angularjs-material.firebaseapp.com/demo/autocomplete. 

I did test the changes with Chromevox and VoiceOver on macOS and they seem to be considerably improved.

**Comment by benjiallen** [comment]
I tested the solution on:

* iOS 11.2.1 + Safari + VoiceOver
* Windows 7 + IE 11 + JAWS 17

## Results for iOS

* Could access the element and begin typing
* Got good feedback on number of results
* Did not get good feedback on field label - would have expected to hear the field labeled as "What is your favorite US state?"
* Could not select a result from the matches found
    * When I close the keyboard the search results disappear

[Video of interaction on Vimeo](https://vimeo.com/289692645)

### Comparison with W3C's ARIA examples

Compared to [WAI-ARIA Authoring Practices for Combobox](https://www.w3.org/TR/wai-aria-practices/#combobox)

This will still hard to use. The big difference with this solution and the solution presented for Angular JS Material is that once the keyboard closes, the results remain visible and I can select a result. This is not intuitive but not impossible to use.

## Results for JAWS

Could not select the element to begin input 🙁

### Comparison with W3C's ARIA examples

Same issue 🙁

## Is there a better way?

This might be a case where the standards are ahead of the technology. A better method might be to provide an alternative to screen reader users. A text field and a group of radio buttons would be a lot easier to get right. The idea being that as you search, you only show radio buttons that match the search term. The screen reader user could then leave the search field and find the radio button that suits their needs, select it and move on.

Not sure what policy the project team has on providing such alternatives. In some cases it can be a lot easier to hide the "complex" component which has poor screen reader support and provide the screen reader an alternative component made up of "simple" components with good support. The sighted user gets the complex component and the simple component is hidden.

This kind of technique is described here [Invisible Content Just for Screen Reader Users](https://webaim.org/techniques/css/invisiblecontent/)

**Comment by Splaktar** [comment]
@benjiallen It looks like JAWS 17 was released in 2015. The WAI-ARIA 1.1 Recommendation came out in December 2017. So that certainly wouldn't be expected to work.

It looks like JAWS has moved to a yearly release with the latest being named "JAWS 2018". It also appears that it is getting approximately monthly releases with improvements to features and bug fixes: https://www.freedomscientific.com/downloads/JAWS/JAWSWhatsNew.

I guess another question here is what versions of screen readers do we support? Only the latest? Or is it expected that some large % of users are stuck on an old version (IE6 flashbacks...)?

**Comment by Splaktar** [comment]
@benjiallen do you know if your VoiceOver iOS results are worse than or an improvement upon 1.1.10 (https://material.angularjs.org/1.1.10/demo/autocomplete)?

**Comment by Splaktar** [comment]
@benjiallen I made some updates to be more compatible with WAI-ARIA 1.0. Hopefully using `role="textbox"` will work better since `role="searchbox"` doesn't exist in WAI-ARIA 1.0.

Can you please re-test against https://angularjs-material.firebaseapp.com/demo/autocomplete?

**Comment by Splaktar** [comment]
@benjiallen I did another round of updates to PR https://github.com/angular/material/pull/11441.

It now should completely align with the [Legacy ARIA 1.0 Combobox With List Autocomplete Example](https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.0pattern/combobox-autocomplete-list.html) instead of the [ARIA 1.1 Combobox with Listbox Popup Example](https://www.w3.org/TR/wai-aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html) as screen reader support for WAI-ARIA 1.1 is very minimal today.

Hopefully this will satisfy JAWS and iOS VoiceOver. Is there any chance that you can run your tests against https://angularjs-material.firebaseapp.com/demo/autocomplete again?

**Comment by Splaktar** [comment]
There were a number of a11y improvements to autocomplete that were merged into `master`. The behavior was verified on macOS VoiceOver, ChromeVox, and to a lesser extent NVDA.

I need someone to test with iOS VoiceOver and JAWS w/ IE11 and report back their findings. If possible, it would be helpful if you could suggest some changes that make JAWS and iOS VoiceOver happier.

You should do these tests against the latest HEAD version: https://material.angularjs.org/HEAD/demo/autocomplete

**Comment by Splaktar** [comment]
### I tested on iOS VoiceOver and found the following
- Typing a char and then swiping right will take you to the clear button and then out of the autocomplete, while the options pop-up stays open
- Touching an item in the options pop-up allows swipe-right and swipe-left to move through the options list
- [ ] Hiding the keyboard dismisses the options pop-up. I need to investigate why this is happening. VoiceOver still puts a border around the autocomplete, but the input loses focus, so the pop-up closes. Perhaps it is focusing the `md-autocomplete` and that focus isn't getting forwarded to the `input`. Need to investigate.
- [ ] Usually the autocomplete is scrolled into view (and out from behind the iOS Virtual keyboard), but sometimes it remains behind the virtual keyboard. In those cases, it is very hard to interact with since you can't see the input or pop-up. Unclear if this is an iOS issue or an issue with our docs site. Needs investigation.

**Comment by Splaktar** [comment]
@jelbourn @benjiallen Now that PR https://github.com/angular/material/pull/11743 is merged, can you please retest this with JAWS?

Please test against HEAD: https://material.angularjs.org/HEAD/demo/autocomplete

**Comment by Splaktar** [comment]
>  Hiding the keyboard dismisses the options pop-up. I need to investigate why this is happening. VoiceOver still puts a border around the autocomplete, but the input loses focus, so the pop-up closes. Perhaps it is focusing the md-autocomplete and that focus isn't getting forwarded to the input. Need to investigate.

This is because dismissing the iOS keyboard sends a `blur` event to the `input`. In every other `blur` case, we want to close the pop-up, but we want to ignore the `blur` event from closing the virtual keyboard on iOS. Ideally, we'd even like to stop the input from being blurred altogether.

**Comment by Splaktar** [comment]
I ran into and opened https://github.com/angular/material/issues/11793 while testing this.

**Comment by Splaktar** [comment]
Comparing a `blur` event from the iOS virtual keyboard closing and from tapping on the backdrop doesn't help as they are identical.

I tried ignoring `blur` events on the input when `$mdUtil.isIos` was true and that works for hiding the virtual keyboard and still being able to interact with the suggestions pop-up. However, it causes issues since it still dismisses the backdrop and detaches the pop-up from the autocomplete (and makes the document body scrollable again). This also means that tapping on the backdrop to dismiss the pop-up doesn't work after you dismiss the virtual keyboard.

**Comment by Splaktar** [comment]
I believe that the final issue that I found related to the virtual keyboard positioning is related to https://github.com/angular/material/issues/11793.

I don't want to pursue the virtual keyboard causing `blur` issue here any further as it appears to be by design on iOS. If that needs to be investigated further, a new issue can be created since it's not specific to VoiceOver on iOS.

I believe that the issues reported here are resolved on [HEAD](https://material.angularjs.org/HEAD/demo/autocomplete) and will be fixed in `1.1.21`.
